keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Multinomial.java

```diff
@@ -0,0 +1,133 @@
+import org.nd4j.linalg.util.ArrayUtil;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import io.improbable.keanu.distributions.DiscreteDistribution;
+import io.improbable.keanu.tensor.Tensor;
+import io.improbable.keanu.tensor.TensorShapeValidation;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.tensor.intgr.IntegerTensor;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+
+
+public class Multinomial implements DiscreteDistribution {
+
+    private final IntegerTensor n;
+    private final DoubleTensor p;
+    private final int numCategories;
+
+    public static Multinomial withParameters(IntegerTensor n, DoubleTensor p) {
+        return new Multinomial(n, p);
+    }
+
+    /**
+     * @see <a href="https://en.wikipedia.org/wiki/Multinomial_distribution">Multinomial Distribution</a>
+     * Generalisation of the Binomial distribution to variables with more than 2 possible values
+     *
+     * @param n The number of draws from the variable
+     * @param p The probability of observing each of the k values (which sum to 1)
+     *          p is a Tensor whose first dimension must be of size k
+     */
+    private Multinomial(IntegerTensor n, DoubleTensor p) {
+        Preconditions.checkArgument(
+            p.sum(0).elementwiseEquals(DoubleTensor.ones(n.getShape())).allTrue(),
+            "Probabilities must sum to one"
+        );
+
+        numCategories = p.getShape()[0];
+        for (int category = 0; category < numCategories; category++) {
+            TensorShapeValidation.checkAllShapesMatch(n.getShape(), p.slice(0, category).getShape());
+        }
+        this.n = n;
+        this.p = p;
+    }
+
+    @Override
+    public IntegerTensor sample(int[] shape, KeanuRandom random) {
+        TensorShapeValidation.checkTensorsMatchNonScalarShapeOrAreScalar(shape, n.getShape());
+
+        Tensor.FlattenedView<Integer> nFlattened = n.getFlattenedView();
+        List<Tensor.FlattenedView<Double>> pFlattened = Lists.newArrayList();
+
+        for (int category = 0; category < numCategories; category++) {
+            DoubleTensor p_i = p.slice(0, category);
+            pFlattened.add(p_i.getFlattenedView());
+        }
+
+        int length = ArrayUtil.prod(shape);
+        int[] samples = new int[0];
+
+        for (int i = 0; i < length; i++) {
+            final int j = i;
+            List<Double> ps = pFlattened.stream().map(p -> p.getOrScalar(j)).collect(Collectors.toList());
+            int[] sample = sample(nFlattened.getOrScalar(i), random, ps.toArray(new Double[0]));
+            samples = ArrayUtils.addAll(samples, sample);
+        }
+        int[] outputShape = shape;
+        if (shape[0] == 1) {
+            outputShape = ArrayUtils.remove(outputShape, 0);
+        }
+        outputShape = ArrayUtils.add(outputShape, numCategories);
+        return IntegerTensor.create(samples, outputShape).transpose();
```

Comment:
I realise now that there was a bug here, which I've just fixed.

Suppose there are k categories. The transpose was necessary because I'm constructing the flat array by repeatedly concatenating k values (which sum to n). That yields me a tensor of shape e.g. [a, b, c, d, k].

The bug is that by taking the transpose I get a tensor of shape [k, d, c, b, a] whereas what I actually need is shape [k, a, b, c, d]. So instead of taking a transpose, I actually need to slice in the last dimension, then concatenate again.

I've written a new method to do this:

    private IntegerTensor constructSampleTensor(int[] shape, int[] samples) {
        int[] outputShape = shape;
        if (shape[0] == 1) {
            outputShape = ArrayUtils.remove(outputShape, 0);
        }
        IntegerTensor abkTensor = IntegerTensor.create(samples, ArrayUtils.add(outputShape, numCategories));
        int[] kabArray = new int[] {};
        for (int category = 0; category < numCategories; category++) {
            IntegerTensor abTensor = abkTensor.slice(outputShape.length, category);
            kabArray = ArrayUtils.addAll(kabArray, abTensor.asFlatIntegerArray());
        }

        return IntegerTensor.create(kabArray, ArrayUtils.insert(0, outputShape, numCategories));
    }